runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/PuzzleSolverOverlay.java

```diff
@@ -230,23 +234,42 @@ else if (config.displayRemainingMoves())
 
 									PuzzleState futureMove = solver.getStep(j);
 
-									if (futureMove == null)
+									if (futureMove == null || !config.drawOverlaps() && positions.contains(futureMove.getEmptyPiece()))
 									{
 										break;
 									}
 
 									int blankX = futureMove.getEmptyPiece() % DIMENSION;
 									int blankY = futureMove.getEmptyPiece() / DIMENSION;
 
-									int markerSize = DOT_MARKER_SIZE - i * 3;
+									int numerator = (i - 1) * (DOT_MARKER_MAX_SIZE - DOT_MARKER_MIN_SIZE);
+									double denominator = movesToShow - 1;
+									int markerSize = (int) Math.round(DOT_MARKER_MAX_SIZE - numerator / denominator);
 
 									int x = puzzleBoxLocation.getX() + blankX * PUZZLE_TILE_SIZE
 											+ PUZZLE_TILE_SIZE / 2 - markerSize / 2;
 
 									int y = puzzleBoxLocation.getY() + blankY * PUZZLE_TILE_SIZE
 											+ PUZZLE_TILE_SIZE / 2 - markerSize / 2;
 
+									Color color;
+									if (config.useDotGradient())
+									{
+										color = ColorUtil.colorLerp(config.dotColor(), config.gradientColor(),
+											(double) (i - 1) / (movesToShow - 1));
+										graphics.setColor(color);
+									}
+									else
+									{
+										color = config.dotColor();
+									}
+									graphics.setColor(color);
 									graphics.fillOval(x, y, markerSize, markerSize);
+
+									graphics.setColor(Color.BLACK);
+									graphics.drawOval(x - 1, y - 1, markerSize + 1, markerSize + 1);
+
+									positions.add(futureMove.getEmptyPiece());
```

Comment:
This line can be removed (or moved up really) if you go with the suggestion above

Reply:
I removed the setColor inside the if statement that wasn't needed. I think everything now is required.